db/system_keyspace.cc

```diff
@@ -1463,7 +1463,47 @@ struct local_cache {
     system_keyspace::bootstrap_state _state;
 };
```

Comment:
> The new function also removes the records with no host_id, so we no longer need the get_host_id function.

I'm not sure we want to do that. It could mask a bug -- such records should never appear in the first place.

Reply:
> Could also be that gossiper was missing HOST ID application state due to a bug (and then we persisted such entry in system.peers) -- and recently we're seeing errors for missing HOST ID, I think they only appear for removed nodes, but it's still unclear why, I believe they should never appear, a node either should have host id attached or not appear at all.

We can check if `host_id` is empty in `system_keyspace::update_peer_info` and trigger `on_internal_error` if it is. I don't know why we don't have such check yet.

My argument was that both of these inconsistencies can occur not only due to node crashes but also due to bugs, so why handle them differently then.

Reply:
> We can check if host_id is empty in system_keyspace::update_peer_info and trigger on_internal_error if it is. I don't know why we don't have such check yet.

Makes sense. @bhalevy please also give your opinion as you were modifying this code recently and probably thought about the problem too

Reply:
It was supposed to be covered in https://github.com/scylladb/scylladb/pull/16376.
There could be use cases to update other fields in system.peers (well the host_id should never change - it really should be the partition key).  So we can require it in `update_peer_info` and do read before write to make sure it's only set once, and in all other cases we can verify it matches the one given by the caller.

Reply:
> well the host_id should never change - it really should be the partition key

It does change -- replace-with-same-IP

Reply:
Yeah, well, it's not the same node anymore.
It's a replacing node that assumes ownership of the same tokens.
But I agree that makes the verification method I suggest impractical, unless we care to delete the previous record (a-la removenode) prior to inserting of the new one.

Reply:
added the check to `system_keyspace::update_peer_info` that `host_id` and `inet_address` are set.